### DISCUSSION TITLE
How to config Server Side in order to catch the positions that its being sent by the plugin

### DISCUSSION THREAD

**Discussion by fabiandepaula** [HaylLtd/cordova-plugin-background-geolocation#125] [Q&A]
The plugin sends a POST to the URL configured. My servier is an Apache with PHP installed and the request never past the first instance that its "W/.expressremise: Got a deoptimization request on un-deoptimizable method void com.google.android.gms.org.conscrypt.NativeCrypto.SSL_do_handshake(long, com.google.android.gms.org.conscrypt.NativeSsl, java.io.FileDescriptor, com.google.android.gms.org.conscrypt.NativeCrypto$SSLHandshakeCallbacks, int)".

Can you help me to do this. Because with JavaScript i dont have any problem to POST anything.

Thanks.

**Comment by HarelM** [comment]
I don't have any experience with this backend server unfortunately...

**Comment by fabiandepaula** [comment]
You are saying i will be the first ti taste this part of the plugin?? But in the example there is a local IP (url: 'http://192.168.81.15:3000/location',). In the documentation there is an item "HTTP Location Posting". Then?

**Reply by HarelM** [reply]
No, I'm saying I'm not familiar with the backend you mentioned.

**Comment by fabiandepaula** [comment]
it is the same as anyone else. Ive never works with java and this kind of connections im not familiar.

**Reply by HarelM** [reply]
It's a basic httpclient written in java...

**Comment by fabiandepaula** [comment]
Ok, but do not work like JS. There is something else in the middle that im not doing well and im asking you to help me, this is my first time working with cordova from inside.

The messages logger by Android:

W/.expressremise: Got a deoptimization request on un-deoptimizable method java.net.InetAddress[] libcore.io.Linux.android_getaddrinfo(java.lang.String, android.system.StructAddrinfo, int)
I/System.out: [socket]:check permission begin!
W/.expressremise: Got a deoptimization request on un-deoptimizable method boolean libcore.io.Linux.access(java.lang.String, int)
W/.expressremise: Got a deoptimization request on un-deoptimizable method boolean libcore.io.Linux.access(java.lang.String, int)
W/.expressremise: Got a deoptimization request on un-deoptimizable method java.lang.Object java.lang.reflect.Method.invoke(java.lang.Object, java.lang.Object[])
W/.expressremise: Got a deoptimization request on un-deoptimizable method void com.google.android.gms.org.conscrypt.NativeCrypto.SSL_do_handshake(long, com.google.android.gms.org.conscrypt.NativeSsl, java.io.FileDescriptor, com.google.android.gms.org.conscrypt.NativeCrypto$SSLHandshakeCallbacks, int)
W/com.marianhello.bgloc.PostLocationTask: Error while posting locations: java.security.cert.CertPathValidatorException: Trust anchor for certification path not found.
W/ActivityRecognitionLocationProvider: Received a Detected Activity that was not tilting / unknown

This is with HTTPS protocol. The next its with the HTTP protocol.

D/com.marianhello.bgloc.PostLocationTask: Executing PostLocationTask#postLocation
D/com.marianhello.bgloc.PostLocationTask: Posting json to url: http://www.caprome.com.ar/express/location headers: {X-FOO=bar}
D/com.marianhello.bgloc.BackgroundGeolocationFacade: Received MSG_ON_STATIONARY
W/com.marianhello.bgloc.PostLocationTask: Error while posting locations: Cleartext HTTP traffic to www.caprome.com.ar not permitted

If i can make this work, i can go forward with my project. Im asking you to help me, because ill never work with JAVA httpclient and ive not have a clou how...

I will try to do my best from my side but i appreciate any help from you. Thanks

**Reply by HarelM** [reply]
I'm not using this part of the plugin, you'll need to find others who did.
I would check if there are headers you need to pass that are done automatically in the js code.
There are issue opened about people using this http part of the plugin. Good luck!

**Reply by fabiandepaula** [reply]
Perfect!! I will be in Alert! And working to try to solve this from my own. Thanks!

**Comment by fabiandepaula** [comment]
Well, y did it!!!

There is a very important question to consider, whether the connection is secure or not ( HTTP or HTTPS ). Because if the connection is secure, you have to make 'ADJUSTMENTS' with the issue of CERTIFICATES, the function after this paragraph solve any problem.

    public SSLContext getTrustContext() {
        TrustManager[] trustAllCerts = new TrustManager[]{
                new X509TrustManager() {
                    public java.security.cert.X509Certificate[] getAcceptedIssuers() {return null;}
                    public void checkClientTrusted(java.security.cert.X509Certificate[] certs, String authType) {};
                    public void checkServerTrusted(java.security.cert.X509Certificate[] certs, String authType) {};
                }
        };

        try {
            SSLContext sc = SSLContext.getInstance("SSL");
            sc.init(null, trustAllCerts, new java.security.SecureRandom());

            return sc;
        } catch (Exception e) {
            e.printStackTrace();

            return null;
        }
    }

As you can see, the function returns a SECURE CONTEXT ( SSLContext ). The next thing to do is bind it to the connection that we are going to make. Something like this.

SSLContext sslContext = getTrustContext();

HttpsURLConnection url = new URL("https://www.example.com").openConnection()

url.setSSLSocketFactory(sslContext.getSocketFactory());

Later I found that the code that sends the connection data did not do so, so I changed it to the following.

url.setRequestProperty("Content-Type", "application/json; charset=utf-8");
url.setRequestProperty("Accept", "application/json");
url.setDoOutput(true);

DataOutputStream wr = new DataOutputStream(url.getOutputStream());

//   data it is the String parameter to send

wr.write(data.getBytes(StandardCharsets.UTF_8));
wr.flush();

int status = url.getResponseCode();

//System.out.println(status);

BufferedReader reader = null;

if (status > 299) {
    reader = new BufferedReader(new InputStreamReader(url.getErrorStream()));
} else {
    reader = new BufferedReader(new InputStreamReader(url.getInputStream()));
}

while ((line = reader.readLine()) != null) {
    buffer.append(line);
}

reader.close();

if (status > 299) {
    throw new IOException( "HttpRequester.sendPosition: " + buffer.toString());
} else {
    rval = new JSONObject(buffer.toString());
}

I think with this code someone can do the code for the other protocol ( HTTP ).

This is the best of me to help to this FANTASTIC project. Thanks!!!

**Comment by jrubiano01** [comment]
I´m working on Codeigniter 3, but I don't know how the data arrives, anyone knows, I send it this way.

url: "BASE_URL/api",
      httpHeaders: {
        'Content-Type': 'application/json'
      },
      postTemplate: {
        id: 1,
        state: 1,
        latitude: '@latitude',
        longitude: '@longitude'
      }

**Comment by fabiandepaula** [comment]
You have to put an ENDPOINT in the property *url*, something like this:
https://yourdomain.com/index.php?c=yourcontroller&m=yourmethod

Then in your method in the variable $_POST['rawdata'] you are going to have
the LOCATION GMS data.

if (isset($_POST['rawdata']) == false AND $raw = '') {
$httpContent = fopen('php://input', 'r');

   while ($kb = fread($httpContent, 1024)) {
    $raw .= $kb;
   }

   fclose($httpContent);

$raw = json_decode($raw) -> rawdata;
}

 I hope this helps. Fabián.
